File: apps/api/app/schemas/feed_monitor.py
"""Schemas for the read-only feed monitor surface."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

FeedMonitorStatus = Literal["ok", "degraded", "down", "unknown", "error"]
FeedMonitorCategory = Literal["feeds_in", "feeds_out", "runtime"]


class FeedMonitorRowSchema(BaseModel):
    id: str
    name: str
    category: FeedMonitorCategory
    kind: str
    status: FeedMonitorStatus
    configured: bool | None = None
    runtime_reachable: bool | None = None
    detail: str | None = None
    action: str | None = None
    checked_at: str | None = None
    latency_ms: float | None = None
    target: str | None = None
    tags: list[str] = Field(default_factory=list)
    extra: dict[str, Any] = Field(default_factory=dict)


class FeedMonitorSummarySchema(BaseModel):
    total: int
    configured: int
    runtime_reachable: int
    issue_count: int
    by_status: dict[str, int] = Field(default_factory=dict)
    by_category: dict[str, int] = Field(default_factory=dict)


class FeedMonitorResponseSchema(BaseModel):
    overall: FeedMonitorStatus
    advisory: str
    as_of_utc: str
    summary: FeedMonitorSummarySchema
    next_actions: list[str] = Field(default_factory=list)
    rows: list[FeedMonitorRowSchema] = Field(default_factory=list)

File: apps/web/lib/api/feedMonitor.ts
import { apiRequest } from "./core";

export type FeedMonitorStatus = "ok" | "degraded" | "down" | "unknown" | "error";
export type FeedMonitorCategory = "feeds_in" | "feeds_out" | "runtime";

export interface FeedMonitorRow {
  id: string;
  name: string;
  category: FeedMonitorCategory;
  kind: string;
  status: FeedMonitorStatus;
  configured: boolean | null;
  runtime_reachable: boolean | null;
  detail: string | null;
  action: string | null;
  checked_at: string | null;
  latency_ms: number | null;
  target: string | null;
  tags: string[];
  extra: Record<string, unknown>;
}

export interface FeedMonitorSummary {
  total: number;
  configured: number;
  runtime_reachable: number;
  issue_count: number;
  by_status: Partial<Record<FeedMonitorStatus, number>>;
  by_category: Partial<Record<FeedMonitorCategory, number>>;
}

export interface FeedMonitorResponse {
  overall: FeedMonitorStatus;
  advisory: string;
  as_of_utc: string;
  summary: FeedMonitorSummary;
  next_actions: string[];
  rows: FeedMonitorRow[];
}

export async function getFeedMonitor(): Promise<FeedMonitorResponse> {
  return apiRequest<FeedMonitorResponse>("/monitor/feeds", { method: "GET" });
}
